third_party/forked/gonum/graph/simple/edgeholder.go

```diff
@@ -16,79 +35,80 @@ type edgeHolder interface {
 	Len() int
 }
 
+// edgeInfo caches the neighbor, avoid interface calls From()/To()/ID()
+type edgeInfo struct {
+	neighbor int
+	edge     graph.Edge
+}
+
 // sliceEdgeHolder holds a list of edges to or from self
 type sliceEdgeHolder struct {
 	self  int
-	edges []graph.Edge
```

Comment:
what if we split this into edgesToSelf and edgesFromSelf? instead of denormalizing data per edge?

Reply:
I'm trying to dig up the benchmarking around this that made me introduce the slice holder in the first place. This is only used for holding edges when we have <= 4 neighbors (since the map overhead dominated in those cases). denormalizing as this PR currently does costs us 8 bytes per edge. splitting lists would match that cost with one neighbor, and win for 2-4 neighbors. 

I also agree with 32 bit node ids, though that was quite a bit more disruptive a change to the upstream library